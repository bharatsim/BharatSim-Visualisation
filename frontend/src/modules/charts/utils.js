import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
