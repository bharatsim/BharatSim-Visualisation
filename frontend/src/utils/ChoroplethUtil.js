import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
