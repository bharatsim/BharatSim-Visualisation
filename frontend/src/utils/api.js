import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
