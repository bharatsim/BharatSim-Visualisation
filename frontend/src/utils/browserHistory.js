// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
