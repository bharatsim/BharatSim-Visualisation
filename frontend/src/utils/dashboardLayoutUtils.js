/* eslint-disable react/prop-types */

import React from 'react';
import renderChart from '../modules/charts/renderChart';

export function renderElement({ layout, chartType, config }) {
  const updatedLayout = { ...layout, y: layout.y ? layout.y : Infinity };
  return (
    <div key={updatedLayout.i} data-grid={updatedLayout} data-testid={updatedLayout.i}>
      <div style={{ height: '100%', width: '100%', boxSizing: 'border-box', padding: '10px' }}>
        {renderChart(chartType, { config })}
      </div>
    </div>
  );
}

export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
