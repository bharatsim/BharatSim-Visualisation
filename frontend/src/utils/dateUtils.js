const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

function getPeriod(hours) {
  return hours >= 12 ? 'PM' : 'AM';
}

function formatDate(dateString) {
  const date = new Date(dateString);
  if (date.toString() === 'Invalid Date') return '--';
  const DD = date.getDate();
  const MM = date.getMonth();
  const YYYY = date.getFullYear();
  const HH = date.getHours();
  const mm = date.getMinutes();

  return `${DD}-${months[MM]}-${YYYY} at ${HH % 12}:${mm < 10 ? `0${mm}` : mm} ${getPeriod(HH)}`;
}

function currentDate(){
  return new Date();
}

export { formatDate, currentDate };
