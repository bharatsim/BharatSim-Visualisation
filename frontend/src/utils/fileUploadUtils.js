import csvParser from 'papaparse';
import dataTypesMapping from '../constants/dataTypesMapping';
import csvParserConfig from '../config/csvParserConfig';
import { loaderStates } from '../hook/useLoader';

function getMessage(fileUploadStatus, fileName) {
  return {
    [loaderStates.ERROR]: `Error occurred while unloading ${fileName}`,
    [loaderStates.SUCCESS]: `${fileName} successfully uploaded`,
    [loaderStates.LOADING]: `uploading ${fileName}`,
  }[fileUploadStatus];
}
function createSchema(row) {
  return Object.keys(row).reduce((acc, element) => {
    acc[element] = dataTypesMapping[typeof row[element]];
    return acc;
  }, {});
}

function parseCsv(csvFile, onComplete, onError) {
  csvParser.parse(csvFile, {
    ...csvParserConfig,
    complete: onComplete,
    error: onError,
  });
}

function resetFileInput(fileInput) {
  // eslint-disable-next-line no-param-reassign
  fileInput.value = '';
  // eslint-disable-next-line no-param-reassign
  fileInput.files = null;
}

function createColumnForMTable(schema) {
  return Object.keys(schema).map((fieldName) => ({
    title: fieldName,
    field: fieldName,
    dataType: schema[fieldName],
  }));
}

export { getMessage, createSchema, parseCsv, resetFileInput, createColumnForMTable };
