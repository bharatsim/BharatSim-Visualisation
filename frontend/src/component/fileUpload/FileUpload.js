import React, { useRef, useState } from 'react';
import { Box, Button } from '@material-ui/core';

import useOldForm from '../../hook/useOldForm';
import FileInput from '../../uiComponent/FileInput';
import useModal from '../../hook/useModal';
import DataTypeConfigModal from './DataTypeConfigModal';

import { api } from '../../utils/api';
import { validateFile } from '../../utils/validators';
import { getMessage, parseCsv, resetFileInput } from '../../utils/fileUploadUtils';
import useLoader, { loaderStates } from '../../hook/useLoader';
import InlineLoader from '../loaderOrError/InlineLoader';

const FILE_INPUT_KEY = 'fileInput';

function FileUpload() {
  const [parsedData, setParsedData] = useState(null);
  const { isOpen, closeModal, openModal } = useModal();
  const {
    loadingState,
    startLoader,
    stopLoaderAfterError,
    stopLoaderAfterSuccess,
    resetLoader,
  } = useLoader();

  const {
    errors,
    validateAndSetValue,
    resetFields,
    onSubmit,
    shouldEnableSubmit,
    values,
  } = useOldForm({
    [FILE_INPUT_KEY]: validateFile,
  });

  const ref = useRef();

  const isUploadDisable = !shouldEnableSubmit() || loadingState.state === loaderStates.LOADING;

  function onCancel() {
    closeModal();
    resetFileInput(ref.current);
  }

  async function upload(file, schema) {
    startLoader(getMessage(loaderStates.LOADING, file.name));
    api
      .uploadFileAndSchema({ file, schema })
      .then(() => {
        stopLoaderAfterSuccess(getMessage(loaderStates.SUCCESS, file.name));
      })
      .catch(() => {
        stopLoaderAfterError(getMessage(loaderStates.ERROR, file.name));
      })
      .finally(() => {
        resetFields([FILE_INPUT_KEY]);
        resetFileInput(ref.current);
      });
  }

  async function onSchemaChangesApply(selectedSchema) {
    await upload(values[FILE_INPUT_KEY], selectedSchema);
    closeModal();
  }

  function validateAndParseCsv() {
    onSubmit((validatedValues) => parseCsv(validatedValues[FILE_INPUT_KEY], onCompleteParseCsv));
  }

  function onCompleteParseCsv(parsedResult) {
    setParsedData(parsedResult.data);
    openModal();
  }

  function onFileInputChange(uploadedFile) {
    resetLoader();
    validateAndSetValue(FILE_INPUT_KEY, uploadedFile);
  }

  return (
    <>
      <Box pb={2}>
        <FileInput
          error={errors[FILE_INPUT_KEY]}
          onChange={onFileInputChange}
          fileUploadStatus={loadingState.state}
          fileUploadStatusMessage={loadingState.message}
          ref={ref}
        />
      </Box>
      <InlineLoader status={loadingState.state} message={loadingState.message} />
      <Button
        type="button"
        onClick={validateAndParseCsv}
        data-testid="button-upload"
        disabled={isUploadDisable}
        variant="contained"
        color="primary"
      >
        Upload
      </Button>
      {isOpen && !!parsedData && (
        <DataTypeConfigModal
          closeModal={closeModal}
          isOpen={isOpen}
          dataRow={parsedData[0]}
          onApply={onSchemaChangesApply}
          onCancel={onCancel}
        />
      )}
    </>
  );
}

export default FileUpload;
