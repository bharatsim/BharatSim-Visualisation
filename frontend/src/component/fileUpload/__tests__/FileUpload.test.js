/* eslint-disable no-console */
import React from 'react';
import { act, fireEvent, render, waitFor, within } from '@testing-library/react';

import FileUpload from '../FileUpload';
import * as fileUtils from '../../../utils/fileUploadUtils';
import { api } from '../../../utils/api';

jest.mock('../../../utils/api', () => ({
  api: {
    uploadFileAndSchema: jest.fn(),
  },
}));

jest.spyOn(fileUtils, 'parseCsv').mockImplementation((csvFile, onComplete) => {
  const data = { data: [{ col1: 'row1', col2: 1 }] };
  onComplete(data);
});

describe('<FileUpload />', () => {
  let renderedComponent;
  let fileInput;
  let uploadButton;

  beforeEach(() => {
    renderedComponent = render(<FileUpload />);
    fileInput = renderedComponent.getByTestId(/input-upload-file/);
    uploadButton = renderedComponent.getByTestId('button-upload');
    api.uploadFileAndSchema.mockResolvedValue('success');
  });

  it('should match a snapshot', () => {
    const { container } = renderedComponent;

    expect(container).toMatchSnapshot();
  });

  it('should enable upload button if csv file is upload', () => {
    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv', size: 2123 }],
      },
    });

    expect(uploadButton).not.toBeDisabled();
  });

  it('should disable upload button if other than csv file is upload', () => {
    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.jpg', type: 'image/jpg' }],
      },
    });

    expect(uploadButton).toBeDisabled();
  });

  it('should show error message if other than csv file is upload', async () => {
    const { getByText } = renderedComponent;

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.jpg', type: 'image/jpg' }],
      },
    });

    const errorText = getByText('Failed to Import file, the format is not supported');
    expect(errorText).toBeInTheDocument();
  });

  it('should not show error message if valid csv file is upload', async () => {
    const { queryByText } = renderedComponent;

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    const errorText = queryByText(
      'Only csv files are allowed of maxmimum size 10MB, Please upload valid csv a file',
    );
    expect(errorText).toBe(null);
  });

  it('should open datatype config modal for selected file on click of upload button', async () => {
    api.uploadFileAndSchema.mockResolvedValue('success');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    await act(async () => {
      fireEvent.click(uploadButton);
    });

    const container = within(document.querySelector('.MuiPaper-root'));
    expect(container.queryByText('Configure Datatype')).toBeInTheDocument();
  });

  it('should upload file and metadata of csv file at upload api', async () => {
    api.uploadFileAndSchema.mockResolvedValue('success');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    fireEvent.click(uploadButton);
    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const applyAndUploadButton = getByText('apply and upload');

    await act(async () => {
      fireEvent.click(applyAndUploadButton);
    });

    expect(api.uploadFileAndSchema).toHaveBeenCalledWith({
      file: {
        name: 'test.csv',
        type: 'text/csv',
      },
      schema: { col1: 'String', col2: 'Number' },
    });
  });

  it('should display uploading message while file is uploading', async () => {
    const { queryByText } = renderedComponent;

    api.uploadFileAndSchema.mockResolvedValue('success');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    fireEvent.click(uploadButton);
    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const applyAndUploadButton = getByText('apply and upload');
    fireEvent.click(applyAndUploadButton);

    const uploading = queryByText('uploading test.csv');
    await waitFor(() => expect(uploading).toBeInTheDocument());
  });

  it('should display success message after file is uploaded', async () => {
    const { queryByText, findByText } = renderedComponent;

    api.uploadFileAndSchema.mockResolvedValue('success');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    fireEvent.click(uploadButton);

    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const applyAndUploadButton = getByText('apply and upload');

    await act(async () => {
      fireEvent.click(applyAndUploadButton);
    });

    await findByText('test.csv successfully uploaded');
    const uploaded = queryByText('test.csv successfully uploaded');
    await waitFor(() => expect(uploaded).toBeInTheDocument());
  });

  it('should display error message after file uploading failed', async () => {
    const { queryByText, findByText } = renderedComponent;

    api.uploadFileAndSchema.mockRejectedValueOnce('failed');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });
    fireEvent.click(uploadButton);

    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const applyAndUploadButton = getByText('apply and upload');

    await act(async () => {
      fireEvent.click(applyAndUploadButton);
    });

    await findByText('Error occurred while unloading test.csv');
    const uploadFiled = queryByText('Error occurred while unloading test.csv');
    await waitFor(() => expect(uploadFiled).toBeInTheDocument());
  });

  it('should reset file upload status after adding new file', async () => {
    const { queryByText, findByText } = renderedComponent;

    api.uploadFileAndSchema.mockRejectedValueOnce('failed');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });
    fireEvent.click(uploadButton);

    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const applyAndUploadButton = getByText('apply and upload');

    await act(async () => {
      fireEvent.click(applyAndUploadButton);
    });

    await findByText('Error occurred while unloading test.csv');
    const uploadFiled = queryByText('Error occurred while unloading test.csv');
    await waitFor(() => expect(uploadFiled).toBeInTheDocument());

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    expect(uploadFiled).not.toBeInTheDocument();
  });

  it('should reset file input after successful upload', async () => {
    api.uploadFileAndSchema.mockResolvedValue('success');

    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    fireEvent.click(uploadButton);
    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const applyAndUploadButton = getByText('apply and upload');

    await act(async () => {
      fireEvent.click(applyAndUploadButton);
    });

    expect(fileInput.files).toBe(null);
    expect(fileInput.value).toBe('');
  });

  it('should reset file input on click of cancel button', async () => {
    fireEvent.change(fileInput, {
      target: {
        files: [{ name: 'test.csv', type: 'text/csv' }],
      },
    });

    fireEvent.click(uploadButton);

    const Modal = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = Modal;

    const cancel = getByText('cancel');

    fireEvent.click(cancel);

    expect(fileInput.files).toBe(null);
    expect(fileInput.value).toBe('');
  });
});
