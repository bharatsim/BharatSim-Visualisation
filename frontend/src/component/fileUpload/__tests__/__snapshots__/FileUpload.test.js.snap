// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileUpload /> should match a snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <input
      accept=".csv"
      data-testid="input-upload-file"
      type="file"
    />
    <div
      class="MuiBox-root MuiBox-root-2"
    />
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
    data-testid="button-upload"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Upload
    </span>
  </button>
</div>
`;
